core/services/relay/evm/event_binding.go

```diff
@@ -326,6 +323,109 @@ func (e *eventBinding) decodeLog(ctx context.Context, log *logpoller.Log, into a
+	sequences := make([]commontypes.Sequence, len(logs))
+
+	for idx := range logs {
+		sequences[idx] = commontypes.Sequence{
+			Cursor: fmt.Sprintf("%s-%s-%d", logs[idx].BlockHash, logs[idx].TxHash, logs[idx].LogIndex),
+			Head: commontypes.Head{
+				Identifier: fmt.Sprint(logs[idx].BlockNumber),
+				Hash:       logs[idx].BlockHash.Bytes(),
+				Timestamp:  uint64(logs[idx].BlockTimestamp.Unix()),
+			},
+		}
+
+		var tpVal reflect.Value
+
+		tpInto := reflect.TypeOf(into)
+		if tpInto.Kind() == reflect.Pointer {
+			tpVal = reflect.New(tpInto.Elem())
+		} else {
+			tpVal = reflect.Indirect(reflect.New(tpInto))
+		}
+
+		// create a new value of the same type as 'into' for the data to be extracted to
+		sequences[idx].Data = tpVal.Interface()
+
+		if err := e.decodeLog(ctx, &logs[idx], sequences[idx].Data); err != nil {
+			return nil, err
+		}
+	}
+
+	return sequences, nil
+}
+
+func (e *eventBinding) remap(filter query.KeyFilter) (query.KeyFilter, error) {
+	remapped := query.KeyFilter{}
+
+	for _, expression := range filter.Expressions {
+		remappedExpression, err := e.remapExpression(filter.Key, expression)
+		if err != nil {
+			return query.KeyFilter{}, err
+		}
+
+		remapped.Expressions = append(remapped.Expressions, remappedExpression)
+	}
+
+	return remapped, nil
+}
+
+func (e *eventBinding) remapExpression(key string, expression query.Expression) (query.Expression, error) {
+	if !expression.IsPrimitive() {
+		remappedBoolExpressions := make([]query.Expression, len(expression.BoolExpression.Expressions))
+
+		for i := range expression.BoolExpression.Expressions {
+			remapped, err := e.remapExpression(key, expression.BoolExpression.Expressions[i])
+			if err != nil {
+				return query.Expression{}, err
+			}
+
+			remappedBoolExpressions[i] = remapped
+		}
+
+		if expression.BoolExpression.BoolOperator == query.AND {
+			return query.And(remappedBoolExpressions...), nil
+		}
+
+		return query.Or(remappedBoolExpressions...), nil
+	}
+
+	return e.remapPrimitive(key, expression)
+}
+
+func (e *eventBinding) remapPrimitive(key string, expression query.Expression) (query.Expression, error) {
+	// remap chain agnostic primitives to chain specific
+	switch primitive := expression.Primitive.(type) {
+	case *primitives.Comparator:
+		if val, ok := e.eventDataWords[primitive.Name]; ok {
+			return logpoller.NewEventByWordFilter(e.hash, val, primitive.ValueComparators), nil
+		}
+
+		return logpoller.NewEventByTopicFilter(e.topics[key].Index, primitive.ValueComparators), nil
+	case *primitives.Confidence:
+		return logpoller.NewConfirmationsFilter(e.confirmationsFrom(primitive.ConfidenceLevel)), nil
+	default:
+		return expression, nil
+	}
+}
+
+func (e *eventBinding) confirmationsFrom(confidence primitives.ConfidenceLevel) int {
+	// sort highest to lowest on confidence
+	sort.Slice(e.confidenceSteps, func(i, j int) bool {
```

Comment:
Does it make sense to sort that slice on every call? We should sort it on-write, right? When `e.confidenceSteps` is initialized you can just sort it once.